### PR TITLE
Spv 92 apm refactoring

### PR DESCRIPTION
**Pull request by andrii-kl**
Screen Shot 2020-04-30 at 12 04 46 PM

### CODE REVIEW

File: nodecore-spv/src/main/java/veriblock/service/AdminApiService.java [hunk 1]
```diff
@@ -45,6 +46,14 @@
 
     VeriBlockMessages.CreateAltChainEndorsementReply createAltChainEndorsement(VeriBlockMessages.CreateAltChainEndorsementRequest request);
 
+    VeriBlockMessages.BlockHeader getLastVBKBlockHeader();
```
Comment (PCasafont) [inline nodecore-spv/src/main/java/veriblock/service/AdminApiService.java:49]: It's preferrable to keep acronyms lowercase for readability (e.g. getLastVbkBlockHeader)

File: nodecore-spv/src/main/java/veriblock/service/impl/Blockchain.java [hunk 2]
```diff
@@ -32,6 +34,7 @@
 
     private final VeriBlockBlock genesisBlock;
     private final VeriBlockStore blockStore;
+    private final EvictingQueue<StoredVeriBlockBlock> blocksCash = EvictingQueue.create(1000);
```
Comment (PCasafont) [inline nodecore-spv/src/main/java/veriblock/service/impl/Blockchain.java:37]: What does "Cash" stand for? Cache?

File: pop-miners/altchain-pop-miner/src/main/kotlin/org/veriblock/lite/NodeCoreLiteKit.kt [hunk 3]
```diff
@@ -98,26 +79,40 @@ class NodeCoreLiteKit(
...
-    fun getAddress(): String = addressManager.defaultAddress.hash
+    fun balanceUpdater() {
+        GlobalScope.launch {
+            while (isActive) {
+                try {
+                    updateBalance()
+                } catch (e: Exception) {
+                    logger.error { e.message; e }
+                }
+                delay(Duration.ofSeconds(60).toMillis())
```
Comment (PCasafont) [inline pop-miners/altchain-pop-miner/src/main/kotlin/org/veriblock/lite/NodeCoreLiteKit.kt:90]:
```suggestion
                delay(60.sec.toMillis())
```

Comment (PCasafont) [reply pop-miners/altchain-pop-miner/src/main/kotlin/org/veriblock/lite/NodeCoreLiteKit.kt:90]: (it needs an import, the IDE will help)

File: pop-miners/altchain-pop-miner/src/main/kotlin/org/veriblock/lite/transactionmonitor/TransactionMonitor.kt [hunk 4]
```diff
@@ -43,122 +49,64 @@ class TransactionMonitor(
...
-        val diskWallet = File(context.directory, context.filePrefix + TM_FILE_EXTENSION)
-
-        lock.withLock {
-            try {
-                FileOutputStream(diskWallet).use { stream ->
-                    stream.writeTransactionMonitor(this@TransactionMonitor)
+    fun start() {
+        GlobalScope.launch {
+            while (isActive) {
+                delay(Duration.ofSeconds(60).toMillis())
```
Comment (PCasafont) [inline pop-miners/altchain-pop-miner/src/main/kotlin/org/veriblock/lite/transactionmonitor/TransactionMonitor.kt:55]:
```suggestion
                delay(60.sec.toMillis()) // or even 1.min.toMillis()!
```

File: pop-miners/altchain-pop-miner/src/main/kotlin/org/veriblock/lite/transactionmonitor/TransactionMonitor.kt [hunk 5]
```diff
@@ -43,122 +49,64 @@ class TransactionMonitor(
...
-                meta.depth = 1
-                meta.addBlockAppearance(block.hash)
-                meta.appearsInBestChainBlock = block.hash
-                meta.appearsAtChainHeight = block.height
-                tx.merklePath = blockTransactions.getValue(tx.id).merklePath
+    private fun checkPendingTransactions() {
+        val pendingTxs = pendingTransactions()
+        val txsInfo = gateway.getTransactions(pendingTxs)
+
+        for (txInfo in txsInfo!!) {
```
Comment (PCasafont) [inline pop-miners/altchain-pop-miner/src/main/kotlin/org/veriblock/lite/transactionmonitor/TransactionMonitor.kt:70]: Careful here, if txsInfo may be null it should be handled properly.
Maybe do something like
```
val txsInfo = gateway.getTransactions(pendingTxs)
    ?: error("Unable to retrieve pending transactions")
```
And then you won't need the non null assertion

File: pop-miners/altchain-pop-miner/src/main/kotlin/org/veriblock/lite/transactionmonitor/TransactionMonitor.kt [hunk 6]
```diff
@@ -43,122 +49,64 @@ class TransactionMonitor(
...
+            if (txInfo.confirmations > MIN_TX_CONFIRMATIONS) {
+                val tx = transactions[Sha256Hash.wrap(txInfo.transaction.txId.toByteArray())]
```
Comment (PCasafont) [inline pop-miners/altchain-pop-miner/src/main/kotlin/org/veriblock/lite/transactionmonitor/TransactionMonitor.kt:72]: Do we expect `tx` to be nullable? If not, throw an error when it's null (like above)

File: pop-miners/altchain-pop-miner/src/main/kotlin/org/veriblock/miners/pop/core/ApmOperation.kt [hunk 7]
```diff
@@ -31,7 +31,10 @@ class ApmOperation(
     logs: List<OperationLog> = emptyList(),
     reconstituting: Boolean = false
 ) : MiningOperation<ApmInstruction, ApmSpTransaction, ApmSpBlock, ApmMerklePath, ApmContext>(
-    id ?: chain.key + UUID.randomUUID().toString().substring(0, 8), endorsedBlockHeight, createdAt, logs, reconstituting
+    id ?: chain.key + UUID.randomUUID().toString().substring(0, 8),
+    endorsedBlockHeight,
+    createdAt,
+    logs, reconstituting
```
Comment (PCasafont) [inline pop-miners/altchain-pop-miner/src/main/kotlin/org/veriblock/miners/pop/core/ApmOperation.kt:37]: Much better, yes

File: pop-miners/altchain-pop-miner/src/main/kotlin/org/veriblock/lite/net/NodeCoreGatewayImpl.kt [hunk 8]
```diff
@@ -256,6 +190,33 @@ class NodeCoreGatewayImpl(
...
+
+    override fun getTransactions(ids: List<Sha256Hash>): List<VeriBlockMessages.TransactionInfo>? {
+        val request =
+            VeriBlockMessages.GetTransactionsRequest.newBuilder()
+                .addAllIds(
+                    ids.stream()
+                        .map { ByteString.copyFrom(it.bytes) }
+                    .collect(Collectors.toList())
+                )
+                .build()
```
Comment (PCasafont) [inline pop-miners/altchain-pop-miner/src/main/kotlin/org/veriblock/lite/net/NodeCoreGatewayImpl.kt:213]:
```suggestion
        val request =VeriBlockMessages.GetTransactionsRequest.newBuilder()
                .addAllIds(ids.map { ByteString.copyFrom(it.bytes) })
                .build()
```

Comment (PCasafont) [reply pop-miners/altchain-pop-miner/src/main/kotlin/org/veriblock/lite/net/NodeCoreGatewayImpl.kt:213]: The kotlin stdlib has some nice inline utiliity extension functions for collections, more optimal and pragmatic for small operations. Also, if you want to use something like java's stream for the middle operation laziness I'd recommend `collection.asSequence()`

File: pop-miners/altchain-pop-miner/src/main/kotlin/org/veriblock/lite/NodeCoreLiteKit.kt [hunk 9]
```diff
@@ -98,26 +79,40 @@ class NodeCoreLiteKit(
...
     }
 
-    fun getAddress(): String = addressManager.defaultAddress.hash
+    fun balanceUpdater() {
+        GlobalScope.launch {
+            while (isActive) {
+                try {
+                    updateBalance()
+                } catch (e: Exception) {
+                    logger.error { e.message; e }
```
Comment (BernatCarbo) [inline pop-miners/altchain-pop-miner/src/main/kotlin/org/veriblock/lite/NodeCoreLiteKit.kt:88]:
```suggestion
                    logger.error(e) { e.message }
```
You can pass the 'e' like that, also the message should be a little more descriptive/graceful than the exception message)

File: pop-miners/altchain-pop-miner/src/main/kotlin/org/veriblock/lite/transactionmonitor/TransactionMonitor.kt [hunk 10]
```diff
@@ -43,122 +49,64 @@ class TransactionMonitor(
...
-                FileOutputStream(diskWallet).use { stream ->
-                    stream.writeTransactionMonitor(this@TransactionMonitor)
+    fun start() {
+        GlobalScope.launch {
+            while (isActive) {
+                delay(Duration.ofSeconds(60).toMillis())
+                try {
+                    checkPendingTransactions()
+                } catch (e: Exception) {
+                    logger.error { e.message; e }
```
Comment (BernatCarbo) [inline pop-miners/altchain-pop-miner/src/main/kotlin/org/veriblock/lite/transactionmonitor/TransactionMonitor.kt:59]:
```suggestion
                    logger.error(e) { e.message; }
```
You can pass the 'e' like that, also the message should be a little more descriptive/graceful than the exception message

**Pull request merged**